File: day16a.lua
#!/usr/bin/env lua

local a = {}

local dx = { 1, 0,-1, 0 }
local dy = { 0, 1, 0,-1 }

local state = {}
local statemt = { __index = state }

function state.new(x, y, d)
    return setmetatable({
        x = x,
        y = y,
        d = d,
    }, statemt)
end

function state:key()
    return self.x | self.y << 8 | self.d << 16
end

function state:clone()
    return state.new(self.x, self.y, self.d)
end

local queue = { r = 1, w = 1 }
local queuemt = { __index = queue }

function queue.new()
    return setmetatable({}, queuemt)
end

function queue:push(x)
    local w = self.w
    self[w] = x
    self.w = w + 1
end

function queue:pop()
    local r = self.r
    assert(r ~= self.w)
    local x = self[r]
    self[r] = nil
    self.r = r + 1
    return x
end

function queue:empty()
    return self.r == self.w
end

local set = {}
local setmt = { __index = set }

function set.new()
    return setmetatable({}, setmt)
end

function set:get(s)
    return self[s:key()] or math.maxinteger
end

function set:set(s, d)
    self[s:key()] = d
end

local function bfs(start, ends)
    local visit = set.new()
    local open = queue.new()
    local shortest = math.maxinteger
    visit:set(start, 0)
    open:push(start)

    while not open:empty() do
        local s = open:pop()
        local dist = visit:get(s)
        --a[s.y][s.x] = 'x'
        if s.x == ends.x and s.y == ends.y then
            if dist < shortest then shortest = dist end
        end

        for dd = -1, 1 do
            local t = s:clone()
            local ndist = dist
            if dd == 0 then
                t.x = s.x + dx[s.d]
                t.y = s.y + dy[s.d]
                ndist = ndist + 1
            else
                t.d = ((s.d - 1 + dd + 4) % 4) + 1
                ndist = ndist + 1000
            end
            if a[t.y][t.x] ~= '#' and ndist < visit:get(t) then
                visit:set(t, ndist)
                open:push(t)
            end
        end
    end

    local count = 0
    for d = 1, 4 do
        local t = ends:clone()
        t.d = d
        if visit:get(t) == shortest then
            open:push(t)
        end
    end
    while not open:empty() do
        local t = open:pop()
        local dist = visit:get(t)
        if a[t.y][t.x] ~= 'O' then
            a[t.y][t.x] = 'O'
            count = count + 1
        end

        for dd = -1, 1 do
            local s = t:clone()
            local ndist = dist
            if dd == 0 then
                s.x = t.x - dx[t.d]
                s.y = t.y - dy[t.d]
                ndist = ndist - 1
            else
                s.d = ((t.d - 1 + dd + 4) % 4) + 1
                ndist = ndist - 1000
            end
            if visit:get(s) == ndist then
                open:push(s)
            end
        end
    end

    return shortest, count
end

local file = io.open("day16.txt")

for line in file:lines() do
    local row = {}
    for i = 1, #line do row[i] = line:sub(i, i) end
    table.insert(a, row)
end

local sx, sy, ex, ey
for y = 1, #a do
    for x = 1, #a[y] do
        if a[y][x] == 'S' then
            a[y][x] = '.'
            sx, sy = x, y
        elseif a[y][x] == 'E' then
            a[y][x] = '.'
            ex, ey = x, y
        end
    end
end
--print("n", #a, "sx", sx, "sy", sy, "ex", ex, "ey", ey)

local sum, sum2 = bfs(state.new(sx, sy, 1), state.new(ex, ey, 1))

for y = 1, #a do
    for x = 1, #a[y] do
        --io.write(a[y][x])
    end
    --print()
end

print(sum, sum2)

File: day16a.c
#include <assert.h>
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

#define COUNT(a) (sizeof(a) / sizeof((a)[0]))

#define WS 8
#define HS 8
#define W (1 << WS)
#define H (1 << HS)
#define A(x, y) a[y][x]
char a[H][W];
int n;

const int dx[] = { 1, 0,-1, 0 };
const int dy[] = { 0, 1, 0,-1 };

typedef union {
    struct {
        unsigned x : WS;
        unsigned y : HS;
        unsigned d : 2;
    };
    unsigned v;
} state_t;

int visit[1 << (WS + HS + 2)];
state_t prev[1 << (WS + HS + 2)][3];

state_t open[1 << 12];
int openrd, openwr;

void push(state_t s)
{
    open[openwr++] = s;
    openwr &= COUNT(open) - 1;
    assert(openrd != openwr);
}

state_t pop()
{
    assert(openrd != openwr);
    state_t s = open[openrd++];
    openrd &= COUNT(open) - 1;
    return s;
}

int dfs(state_t t)
{
    int sum = 0;
    if (A(t.x, t.y) != 'O') {
        A(t.x, t.y) = 'O';
        sum++;
    }
    for (int p = 0; p < 3; p++) {
        state_t s = prev[t.v][p];
        if (s.v) {
            prev[t.v][p].v = 0;
            sum += dfs(s);
        }
    }
    return sum;
}

void insert(state_t s, state_t t, int ndist)
{
    if (ndist > visit[t.v]) return;
    if (ndist < visit[t.v]) memset(prev[t.v], 0, sizeof(prev[t.v]));
    for (int p = 0; p < 3; p++) {
        if (prev[t.v][p].v == s.v) return;
        if (!prev[t.v][p].v) {
            prev[t.v][p] = s;
            break;
        }
    }
    visit[t.v] = ndist;
    push(t);
}

int bfs(state_t start, state_t end, int* count)
{
    for (int i = 0; i < COUNT(visit); i++) {
        visit[i] = INT_MAX;
    }
    int shortest = INT_MAX;
    visit[start.v] = 0;
    push(start);

    while (openwr != openrd) {
        state_t s = pop();
        int dist = visit[s.v];
        //A(s.x, s.y) = 'x';
        if (s.x == end.x && s.y == end.y) {
            if (dist < shortest) shortest = dist;
        }

        int nx = (int)s.x + dx[s.d];
        int ny = (int)s.y + dy[s.d];
        if (A(nx, ny) != '#') {
            state_t t = s;
            t.x = nx;
            t.y = ny;
            insert(s, t, dist + 1);
        }

        for (int dd = -1; dd <= 1; dd += 2) {
            state_t t = s;
            t.d = (s.d + dd + 4) % 4;
            insert(s, t, dist + 1000);
        }
    }

    *count = 0;
    for (int d = 0; d < 4; d++) {
        state_t t = end;
        t.d = d;
        if (visit[t.v] == shortest) {
            *count += dfs(t);
        }
    }
    return shortest;
}

int main()
{
    FILE* file = fopen("day16.txt", "r");
    int sum = 0, sum2 = 0;

    while (1 == fscanf(file, "%s", &A(0, n))) n++;

    int sx, sy, ex, ey;
    for (int y = 0; y < n; y++) {
        for (int x = 0; x < n; x++) {
            if (A(x, y) == 'S') {
                A(x, y) = '.';
                sx = x, sy = y;
            } else if (A(x, y) == 'E') {
                A(x, y) = '.';
                ex = x, ey = y;
            }
        }
    }
    //printf("n %d sx %d sy %d ex %d ey %d\n", n, sx, sy, ex, ey);

    sum = bfs((state_t){{ sx, sy, 0 }}, (state_t){{ ex, ey, 0 }}, &sum2);

    for (int y = 0; y < n; y++) {
        for (int x = 0; x < n; x++) {
            //printf("%c", A(x, y));
        }
        //printf("\n");
    }

    printf("%d %d\n", sum, sum2);
}
